### PR TITLE
Cleanup prescriptor/NN, move mutation

### PR DESCRIPTION
**Pull request by danyoungday**
Made the prescriptor factory and NN implementations nicer. Moved mutation out of crossover and into its own function

### CODE REVIEW

File: presp/evolution.py [hunk 1]
```diff
@@ -25,12 +25,28 @@ def __init__(self,
                  mutation_rate: float,
                  mutation_factor: float,
                  save_path: str,
-                 seed_dir: str,
                  prescriptor_factory: PrescriptorFactory,
                  evaluator: Evaluator,
+                 seed_dir: str = None,
```
Comment (danyoungday) [inline presp/evolution.py:30]: Made seed dir an optional param instead of requiring it be filled with `None`

File: presp/evolution.py [hunk 2]
```diff
@@ -100,8 +116,17 @@ def create_pop(self, population: list[Prescriptor], remove_population_pct: float
         next_pop = []
         n_children = self.population_size - self.n_elites
         while len(next_pop) < n_children:
+            # Select, crossover, mutate
             parents = self.selection(top_pop)
-            children = self.prescriptor_factory.crossover(parents, self.mutation_rate, self.mutation_factor)
+            children = self.prescriptor_factory.crossover(parents)
+            for child in children:
+                self.prescriptor_factory.mutation(child, self.mutation_rate, self.mutation_factor)
```
Comment (danyoungday) [inline presp/evolution.py:123]: Now we mutate outside of crossover so that we can have separation of duties

File: presp/evolution.py [hunk 3]
```diff
@@ -100,8 +116,17 @@ def create_pop(self, population: list[Prescriptor], remove_population_pct: float
...
+
+            # Tag children with parent ids
+            parent_ids = [parent.cand_id for parent in parents]
+            for child in children:
+                child.parents = parent_ids
```
Comment (danyoungday) [inline presp/evolution.py:128]: Parent tagging

File: presp/evolution.py [hunk 4]
```diff
@@ -128,9 +153,14 @@ def record_results(self):
             # Save to file if it's new. Only save rank 1 candidates if save_all is False.
             cand_gen = int(candidate.cand_id.split("_")[0])
             if (candidate.rank == 1 or self.save_all) and cand_gen == self.generation:
-                candidate.save(self.save_path / str(cand_gen) / f"{candidate.cand_id}")
+                self.prescriptor_factory.save(candidate, (self.save_path / str(cand_gen) / f"{candidate.cand_id}"))
```
Comment (danyoungday) [inline presp/evolution.py:156]: Factory is in charge of saving now

File: presp/prescriptor/base.py [hunk 5]
```diff
@@ -12,6 +11,7 @@ class Prescriptor(ABC):
     """
     def __init__(self):
         self.cand_id = ""
+        self.parents = None
         self.metrics = None
         self.outcomes = None
         self.rank = None
```
Comment (danyoungday) [inline presp/prescriptor/base.py:17]: A new parameter for `Prescriptor` to maintain

File: presp/prescriptor/base.py [hunk 6]
```diff
@@ -24,42 +24,46 @@ def forward(self, context):
...
-    Abstract class in charge of creating prescriptors.
-    Takes in a prescriptor class (not instance) and creates prescriptors from it.
-    Implementations should store details used to create prescriptors.
+    Interface in charge of creating prescriptors.
+    Prescriptors should be able to be randomly initialized at the start of evolution, crossed over and mutated,
+    and saved and loaded.
     """
-    def __init__(self, prescriptor_cls: Type[Prescriptor]):
-        self.prescriptor_cls = prescriptor_cls
-
```
Comment (danyoungday) [inline presp/prescriptor/base.py:46]: Remove this clunkly CLS stuff. We can just implement a new `PrescriptorFactory` for every `Prescriptor`

File: presp/prescriptor/base.py [hunk 7]
```diff
@@ -24,42 +24,46 @@ def forward(self, context):
...
+        """
+
+    @abstractmethod
+    def mutation(self, candidate: Prescriptor, mutation_rate: float, mutation_factor: float):
+        """
+        Mutates a prescriptor in-place.
+        :param candidate: The candidate to mutate.
         :param mutation_rate: The rate at which to mutate each parameter.
         :param mutation_factor: The factor by which to mutate each parameter.
         """
```
Comment (danyoungday) [inline presp/prescriptor/base.py:57]: Mutation is pulled out of crossover

File: presp/prescriptor/base.py [hunk 8]
```diff
@@ -24,42 +24,46 @@ def forward(self, context):
...
 
+    @abstractmethod
+    def save(self, candidate: Prescriptor, path: Path):
+        """
+        Save the prescriptor to file.
+        :param prescriptor: The prescriptor to save.
+        :param path: The path to save the prescriptor to.
+        """
+
```
Comment (danyoungday) [inline presp/prescriptor/base.py:66]: Saving is pulled out of `Prescriptor` and into `PrescriptorFactory`

File: presp/prescriptor/nn.py [hunk 9]
```diff
@@ -14,33 +13,43 @@ class NNPrescriptor(Prescriptor):
     """
     Simple neural network implementation of a prescriptor.
     """
-    def __init__(self, model_params: dict[str, int], device: str = "cpu"):
+    def __init__(self, model_params: list[dict], device: str = "cpu"):
         super().__init__()
-        self.model_params = model_params
+        self.model_params = [{**layer} for layer in model_params]
```
Comment (danyoungday) [inline presp/prescriptor/nn.py:18]: Deep copy model params so we don't mess them up when constructing the model

File: presp/prescriptor/nn.py [hunk 10]
```diff
@@ -14,33 +13,45 @@ class NNPrescriptor(Prescriptor):
...
+            if layer_type == "linear":
+                layers.append(torch.nn.Linear(**layer))
+            elif layer_type == "tanh":
+                layers.append(torch.nn.Tanh(**layer))
+            elif layer_type == "sigmoid":
+                layers.append(torch.nn.Sigmoid(**layer))
+            elif layer_type == "softmax":
+                layers.append(torch.nn.Softmax(**layer))
+            else:
+                raise ValueError(f"Unknown layer type: {layer_type}")
```
Comment (danyoungday) [inline presp/prescriptor/nn.py:34]: We can add more layer types later but for now this is all we need

File: presp/prescriptor/nn.py [hunk 11]
```diff
@@ -14,33 +13,45 @@ class NNPrescriptor(Prescriptor):
...
-        return self.model(context)
-
-    def save(self, path: Path):
-        torch.save(self.model.state_dict(), path)
+        """
+        We simply pass the context through our model here. The user should get exactly out of the prescriptor what
+        they created using the config. They can process the output as they see fit.
+        """
+        output = self.model(context)
+        return output
```
Comment (danyoungday) [inline presp/prescriptor/nn.py:47]: Forward should just do a forward pass of the defined model so as not to confuse the user. We can modify the output ourselves in the evaluator.

File: presp/prescriptor/nn.py [hunk 12]
```diff
@@ -50,38 +61,33 @@ def random_init(self) -> NNPrescriptor:
...
         """
-        candidate = self.prescriptor_cls(self.model_params, self.device)
-        for i, layer in enumerate(candidate.model):
+        candidate = NNPrescriptor(self.model_params, self.device)
+        linear_layers = [layer for layer in candidate.model if isinstance(layer, torch.nn.Linear)]
+        for i, layer in enumerate(linear_layers):
             if isinstance(layer, torch.nn.Linear):
-                # TODO: This relies on the model ending with a linear layer which we may not guarantee later!
-                if i != len(candidate.model) - 1:
+                if i != len(linear_layers) - 1:
```
Comment (danyoungday) [inline presp/prescriptor/nn.py:68]: We have to specify these are linear layers now
